File: LeetCode/110/main.py
"""
DIFFICULTY : easy
TAGS : tree, dfs, binary tree
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        def check_diff(root) :
            if(root is None) :
                return 0

            ls = check_diff(root.left)
            rs = check_diff(root.right)
            if(ls == -1 or rs == -1 or abs(ls - rs) > 1) :
                return -1
            
            return 1 + max(ls, rs)
        
        return check_diff(root) != -1        

File: LeetCode/110/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  int maxDepth(TreeNode *node, bool &balanced) {
    if (!node) {
      return 0;
    }
    int leftHeight = maxDepth(node->left, balanced);
    int rightHeight = maxDepth(node->right, balanced);
    if (abs(leftHeight - rightHeight) > 1) {
      balanced = false;
    }
    return 1 + max(leftHeight, rightHeight);
  }

  bool isBalanced(TreeNode *root) {
    bool ans = true;
    maxDepth(root, ans);
    return ans;
  }
};